### DISCUSSION TITLE
set compilation_mode in cc_binary rule

### DISCUSSION THREAD

**Discussion by sonicxml** [bazelbuild/bazel#20089] [Q&A] [answered]
Hello,

In my bazel setup, I am building docker images that contain a binary produced by a `cc_binary` rule plus some configuration files, e.g.
```
cc_binary(
  name = "foo",
  srcs = ["src/main.cpp"],
  deps = [
    "//common:common",
  ],
  visibility = ["//visibility:public"],
)
container_image(
  name = "foo_image",
  base = "//:cpp_base_image",
  files = [
    "//services/foo:foo",
  ] + glob(["config/*"]),
)
```

However, I would like the image rules to be able to choose whether they want a debug or release version of the binary and to have this work independently of any `-c` flags set on the command line. What is the best way of going about this?

My thought is to define two custom rules that wrap `cc_binary`, e.g.
```
debug_cc_binary = rule(
  compilation_mode = "debug",
)
release_cc_binary = rule(
  compilation_mode = "opt",
)
```
But I'm not quite sure how to go about doing that either. Any advice would be appreciated, thanks!

**Comment by keith** [comment, marked as answer]
You can do this with transitions, there is a bit of a guide on how to do this in https://bazel.build/extending/config. Here's an example you can copy doing this too https://github.com/apple/swift-syntax/blob/098b177830c0d2e614b3757241b2572d758f9143/utils/bazel/opt_wrapper.bzl

**Reply by fmeum** [reply]
If you want a quick start, you can also use `with_cfg.bzl`, which has an example of an `opt_filegroup`: https://github.com/fmeum/with_cfg.bzl/blob/main/examples/opt_filegroup/opt_filegroup.bzl

**Reply by sonicxml** [reply]
Thanks! I was able to do it with transitions

**Comment by tpudlik** [comment]
Isn't a really simple solution to have two different `cc_binary` targets,
with different `copts` attributes? So,

cc_binary(
  name = "foo_opt",

  copts = ["opt"],

  srcs = ["src/main.cpp"],

  deps = [
    "//common:common",
  ],
  visibility = ["//visibility:public"],
)


cc_binary(
  name = "foo_dbg",

  copts = ["dbg"],

  srcs = ["src/main.cpp"],

  deps = [
    "//common:common",
  ],
  visibility = ["//visibility:public"],
)

**Comment by tpudlik** [comment]
Ah I guess that's too simple, since the `copts` will not propagate down to
the dependencies.
